### PR TITLE
Migrate to Typescript

### PR DESCRIPTION
**Pull request by s1na**
Breaking changes are how classes are exported (using named exports). Users can import one of `CheckpointTrie, SecureTrie, BaseTrie` via `import { SecureTrie } from 'merkle-patricia-tree'.

I don't consider this migration complete. The library in its current is difficult to type and I had to use a lot of type casting (e.g. `key as Buffer`) and ignoring TS errors via `// @ts-ignore` for now to get it to compile, and to avoid making this PR bigger than it is. Some parts really need refactoring to have proper types (specially `TrieNode`), which I plan to do in next PRs. Promisification is also left for future PRs.

I've tested this branch in `ethereumjs-vm` and tests are passing

### CODE REVIEW

**Comment by holgerd77** [comment]
@s1na Have updated this branch with the Travis xvfb fix, now the main tests generally pass but there seems to be some babel version related problem left preventing the `coveralls` and `browser` test runs to pass, can you please fix?

**Comment by alcuadrado** [comment]
I haven't reviewed the code but got curious about the xvfb thing, so I tried running the tests locally and I'm getting a different set of errors. I think there's no need for babel here, so I copied the `karma.conf.js` from the [`-block` TS migration PR](https://github.com/ethereumjs/ethereumjs-block/pull/72), and it works with some minor changes.

It's here:

```js

module.exports = function(config) {
  config.set({
    browserNoActivityTimeout: 60000,
    frameworks: ['browserify', 'detectBrowsers', 'tap'],
    files: ['./test/*.js'],
    preprocessors: {
      './dist/**/*.js': ['browserify'],
      './test/**/*.js': ['browserify']
    },
    singleRun: true,
    detectBrowsers: {
      enabled: true,
      usePhantomJS: false,
      postDetection: function(availableBrowsers) {
        return ['Firefox']
      },
    }
  })
}
```

This changes also let us remove the file `.babelrc` and these dependencies: `babelify`, `karma-chrome-launcher`, and `karma-env-preprocessor`.

**Comment by holgerd77** [comment]
Ok, tests working, seems coverage run needs some additional fix?

**Comment by s1na** [comment]
I think it might have something with the fact that we run tests on `dist` and not on `src`. But even then, `coveralls` should have run and gave weird results.

Another point is that when I run `npm run coverage` locally, it doesn't produce `.nyc_outpu/lcov.info`, but when I run `npx nyc npm run test && npx nyc report --reporter=text-lcov > .nyc_output/lcov.info` explicitly it does. Not sure yet what's the cause.

**Comment by s1na** [comment]
By the way, thanks @alcuadrado for the fixed karma config :)

**Comment by s1na** [comment]
Modified `coverage`. It now does generate coverage for `dist` files, and does produce a `lcov.info`. But the coveralls task still seems to be stuck. Out of ideas, if anyone else has any idea I could try them out.

**Comment by alcuadrado** [comment]
(closing and reopening this PR, as that supposedly fixes coveralls)

**Comment by alcuadrado** [comment]
That didn't work. We should try [this](https://github.com/lemurheavy/coveralls-public/issues/905#issuecomment-271040338), but I don't have enough permissions. @holgerd77 maybe?

**Comment by s1na** [comment]
Finally! Either updating the `coveralls` package or using `npm run coverage && coveralls < .nyc_output/lcov.info` in the script instead of the ethereumjs-config coveralls script seems to have worked (I'm guessing the first one).

Coverage went down 2%, probably because it's measuring coverage over the compiled files and not the sources. Not sure what to do about it

Before this, I also added CircleCI config files (to test if `coveralls` works there). Do we want to switch to circle or remain with travis?

**Comment by holgerd77** [comment]
Great! 😄

I would have a tendency to not side-introduce CI tool switches here and keep on Travis at least along with this PR.

**Comment by holgerd77** [comment]
@s1na Have increased the failure decrease threshold, for the TS transitions I think it's ok to make an exception with all these heavy structural changes likely shifting things a bit.

There is still a 90% absolute threshold which will cover us here.

**Comment by s1na** [comment]
Ok, removed the circleci config, but only afterwards removed this repo from the circleci app, that's why it tried to build but failed.

If necessary I can do a rebase after the review to "fix" the circleci error.

**Comment by holgerd77** [comment]
Yes, maybe a rebase would be good here, especially for these large/central PRs it looks a bit ugly if CI is displayed as "failing". Can you actually do before the review, otherwise the approval will be again dismissed on rebase?

File: .gitignore [hunk 1]
```diff
@@ -1,16 +1,5 @@
...
-/db.js
-/index.js
-/prioritizedTaskExecutor.js
-/proof.js
-/readStream.js
-/scratch.js
-/scratchReadStream.js
-/secure.js
-/trieNode.js
-/util/
```
Comment (holgerd77) [inline .gitignore]: Phew, nice to have this finally out here. 👍 😛

File: .nycrc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@ethereumjs/config-nyc",
+  "include": [
+    "dist/**/*.js"
+  ],
+  "extension": [
+    ".js"
+  ],
+  "source-map": false
+}
```
Comment (holgerd77) [inline .nycrc]: Side note: we should from time to time re-evaluate a bit which settings which are manually added to the config files make generally sense and can be shifted to the generic config files.

File: package.json [hunk 3]
```diff
@@ -40,41 +50,35 @@
...
+    "prettier": "^1.18.2",
+    "tape": "^4.10.1",
+    "tslint": "^5.18.0",
+    "typedoc": "^0.15.0",
+    "typedoc-plugin-markdown": "^2.1.0",
+    "typescript": "^3.5.3",
+    "typestrict": "^1.0.2"
   },
   "contributors": [
     "Aaron Kumavis <http://aaron.kumavis.me/> (https://github.com/kumavis)"
```
Comment (holgerd77) [inline package.json]: Ok (whole file).

**Comment by holgerd77** [comment]
Update: or maybe wait, I'll likely have some small comments to update.

File: README.md [hunk 4]
```diff
@@ -99,20 +102,23 @@ trie.get(address, function (err, data) {
...
+    })
+    .on('end', function() {
+      console.log('Finished reading storage.')
+    })
 })
 ```
 
 # API
+
 [./docs/](./docs/index.md)
```
Comment (holgerd77) [inline README.md]: Did you generally leave out doc generation on purpose to keep this on the released version?

**Comment by s1na** [comment]
My main concern for rebasing was that Github might not detect modifications to the files and doesn't show a good diff, which would make the review process harder.

File: README.md (hunk 4 above)
Comment (s1na) [reply README.md]: Yeah, I remember something like this discussed somewhere :) plus they will most likely change in the next PRs

File: src/checkpointTrie.ts [hunk 5]
```diff
@@ -147,10 +150,37 @@ module.exports = class CheckpointTrie extends BaseTrie {
...
+          key: hashRoot,
+          value: rlpNode,
+        })
+      }
+
+      return hashRoot
+    }
+
+    return node.raw
+  }
```
Comment (holgerd77) [inline src/checkpointTrie.ts]: Ah, just seeing this which is likely related to my `_formatNode` question in `baseTrie`, could you nevertheless give a short explanation of the effects here? Does this fixes a bug? Make things more secure?

File: src/checkpointTrie.ts [hunk 6]
```diff
@@ -147,10 +150,37 @@ module.exports = class CheckpointTrie extends BaseTrie {
...
 }
```
Comment (holgerd77) [inline src/checkpointTrie.ts]: Ok (whole file).

File: src/db.ts [hunk 7]
```diff
@@ -77,7 +85,7 @@ module.exports = class DB {
    * Returns a copy of the DB instance, with a reference
    * to the **same** underlying leveldb instance.
    */
-  copy () {
+  copy(): DB {
     return new DB(this._leveldb)
   }
 }
```
Comment (holgerd77) [inline src/db.ts]: Ok (whole file, `db.ts`).

File: src/prioritizedTaskExecutor.ts [hunk 8]
```diff
@@ -20,21 +29,21 @@ module.exports = class PrioritizedTaskExecutor {
    * @param {Number} priority The priority of the task
    * @param {Function} task The function that accepts the callback, which must be called upon the task completion.
```
Comment (holgerd77) [inline src/prioritizedTaskExecutor.ts]: `fn` has to be renamed in the doc comment as well.

File: src/prioritizedTaskExecutor.ts [hunk 9]
```diff
@@ -20,21 +29,21 @@ module.exports = class PrioritizedTaskExecutor {
...
         }
       })
     } else {
       this.queue.push({
         priority: priority,
-        task: task
+        fn: fn,
       })
     }
   }
```
Comment (holgerd77) [inline src/prioritizedTaskExecutor.ts]: Ok (whole file, `prioritizedTaskExecutor.ts`)

File: src/readStream.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+          next()
+        },
+        () => {
+          // close stream
+          this.push(null)
+        },
+      )
+    }
+  }
+}
```
Comment (holgerd77) [inline src/readStream.ts]: Ok (whole file, `readStream.ts`).

File: src/scratch.ts [hunk 11]
```diff
@@ -34,7 +36,7 @@ module.exports = class ScratchDB extends DB {
     asyncFirstSeries(getDBs, dbGet, cb)
   }
 
-  copy () {
+  copy(): ScratchDB {
     const scratch = new ScratchDB(this._upstream)
     scratch._leveldb = this._leveldb
     return scratch
```
Comment (holgerd77) [inline src/scratch.ts]: Ok (whole file, `scratch.ts`).

File: src/scratchReadStream.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+          next()
+        },
+        () => {
+          // close stream
+          this.push(null)
+        },
+      )
+    }
+  }
+}
```
Comment (holgerd77) [inline src/scratchReadStream.ts]: Ok (whole file, `scratchReadStream.ts`).

File: src/secure.ts [hunk 13]
```diff
@@ -9,47 +9,47 @@ const CheckpointTrie = require('./checkpointTrie')
...
     }
   }
 
-  del (key, cb) {
-    const hash = ethUtil.keccak256(key)
+  del(key: Buffer, cb: Function) {
+    const hash = keccak256(key)
     super.del(hash, cb)
   }
 }
```
Comment (holgerd77) [inline src/secure.ts]: Ok (whole file, `secure.ts`).

File: src/trieNode.ts [hunk 14]
```diff
@@ -1,20 +1,31 @@
...
+import { keccak256 } from 'ethereumjs-util'
+import { stringToNibbles, nibblesToBuffer } from './util/nibbles'
+import { isTerminator, addHexPrefix, removeHexPrefix } from './util/hex'
+
+enum NodeType {
+  Leaf = 'leaf',
+  Branch = 'branch',
+  Extension = 'extention', // FIXME
+  Unknown = 'unknown', // TODO
+}
```
Comment (holgerd77) [inline src/trieNode.ts]: Nice.

File: src/trieNode.ts [hunk 15]
```diff
@@ -35,63 +46,66 @@ module.exports = class TrieNode {
...
-  set key (k) {
+  set key(k) {
     this.setKey(k)
   }
 
-  parseNode (rawNode) {
+  parseNode(rawNode: Buffer[]) {
     this.raw = rawNode
     this.type = TrieNode.getNodeType(rawNode)
   }
```
Comment (holgerd77) [inline src/trieNode.ts]: Can `parseNode` now be deprecated or removed?

File: src/trieNode.ts [hunk 16]
```diff
@@ -1,20 +1,31 @@
...
+      this.type = TrieNode.getNodeType(type)
     } else {
       this.type = type
       if (type === 'branch') {
         var values = key
+        // @ts-ignore
         this.raw = Array.apply(null, Array(17))
         if (values) {
-          values.forEach(function (keyVal) {
+          values.forEach(function(this: any, keyVal: any) {
```
Comment (holgerd77) [inline src/trieNode.ts]: Can't really read this code TBH, was the missing `this` a bug before or is this a structural adoption?

File: src/trieNode.ts [hunk 17]
```diff
@@ -157,12 +173,10 @@ module.exports = class TrieNode {
...
+        for (let index = 0, end = 16; index < end; index++) {
+          const value = this.getValue(index)
           if (value) {
-            children.push([
-              [index], value
-            ])
+            children.push([[index], value])
           }
         }
         break
```
Comment (holgerd77) [inline src/trieNode.ts]: Looks good (whole file, `trieNode.ts`), some comments.

File: src/util/async.ts [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+        next(err)
+      })
+    },
+    function() {
+      if (!didComplete) {
+        cb()
+      }
+    },
+  )
+}
```
Comment (holgerd77) [inline src/util/async.ts]: Whew, this code is so hard to decipher (just a note): but changes ok.

File: src/util/hex.ts [hunk 19]
```diff
@@ -43,12 +43,6 @@ function removeHexPrefix (val) {
...
-function isTerminator (key) {
+export function isTerminator(key: number[]): boolean {
   return key[0] > 1
 }
-
-module.exports = {
-  addHexPrefix,
-  removeHexPrefix,
-  isTerminator
-}
```
Comment (holgerd77) [inline src/util/hex.ts]: Ok (whole file, `util/hex.js`).

File: src/util/nibbles.ts [hunk 20]
```diff
@@ -53,14 +53,7 @@ function matchingNibbleLength (nib1, nib2) {
...
   const length = matchingNibbleLength(keyA, keyB)
   return length === keyA.length && length === keyB.length
 }
-
-module.exports = {
-  stringToNibbles,
-  nibblesToBuffer,
-  matchingNibbleLength,
-  doKeysMatch
-}
```
Comment (holgerd77) [inline src/util/nibbles.ts]: Ok (whole file, `util/nibbles.ts`).

File: test/streams.js [hunk 21]
```diff
@@ -1,4 +1,4 @@
-const Trie = require('../src/index.js')
+const Trie = require('../dist/index.js').CheckpointTrie
 const describe = require('tape')
 
 describe('kv stream test', function (tester) {
```
Comment (holgerd77) [inline test/streams.js]: Can we have the `require` statements consistent without the `.js` appendix - like e.g. `const ScratchDB = require('../dist/scratch').ScratchDB`? Otherwise test file changes look all ok.

File: src/baseTrie.ts [hunk 22]
```diff
@@ -271,51 +308,60 @@ module.exports = class Trie {
...
   /*
    * Finds all nodes that store k,v values
    */
-  _findNode (key, root, stack, cb) {
+  /*_findNode (key: Buffer, root: Buffer, stack, cb) {
     this.findPath(key, () => {
       cb.apply(null, arguments)
     })
-  }
+  }*/
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: Oh sorry forgot to do something about that. Also did an organization-wide search and we're not using it anywhere. The implementation also seems off (e.g. `stack` param not being used at all). I deleted it but we can revive it from git history if necessary.

File: src/baseTrie.ts [hunk 23]
```diff
@@ -328,12 +374,15 @@ module.exports = class Trie {
...
    * @param {Function} cb - the callback
    */
-  _updateNode (key, value, keyRemainder, stack, cb) {
-    const toSave = []
+  _updateNode(k: Buffer, value: Buffer, keyRemainder: number[], stack: TrieNode[], cb: Function) {
+    const toSave: BatchDBOp[] = []
     const lastNode = stack.pop()
+    if (!lastNode) {
+      throw new Error('Stack underflow')
+    }
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: We'll probably need to introduce exceptions in a few more places.

File: src/baseTrie.ts [hunk 24]
```diff
@@ -389,16 +441,16 @@ module.exports = class Trie {
       stack.push(newBranchNode)
 
       if (lastKey.length !== 0) {
-        const branchKey = lastKey.shift()
+        const branchKey = lastKey.shift() as number
 
         if (lastKey.length !== 0 || lastNode.type === 'leaf') {
           // shriking extention or leaf
           lastNode.key = lastKey
           const formatedNode = this._formatNode(lastNode, false, toSave)
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: Yeah I changed the signature because it was both confusing and didn't play nice with typescript. The reason was that out of the 4 arguments, the 3rd one was optional and defaulted to `false`. I moved the optional argument to be the last one. That's why this line didn't need to change. It wasn't a bug before.

File: src/baseTrie.ts [hunk 25]
```diff
@@ -603,36 +666,37 @@ module.exports = class Trie {
...
+    let lastNode = stack.pop() as TrieNode
+    assert(lastNode)
     let parentNode = stack.pop()
-    const opStack = []
+    const opStack: BatchDBOp[] = []
 
-    if (!Array.isArray(key)) {
-      // convert key to nibbles
-      key = stringToNibbles(key)
-    }
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: It was checking if the key has been converted to `nibbles` yet or not. So it was supporting the key in two formats. I opted for explicitly supporting only one format which the `del` method provides to `_deleteNode`.

File: src/baseTrie.ts [hunk 26]
```diff
@@ -649,11 +713,11 @@ module.exports = class Trie {
         const branchNodeKey = branchNodes[0][0]
 
         // look up node
-        this._lookupNode(branchNode, (e, foundNode) => {
+        this._lookupNode(branchNode, (e: Error, foundNode: TrieNode) => {
           if (e) {
             return cb(e, foundNode)
           }
-          key = processBranchNode(key, branchNodeKey, foundNode, parentNode, stack, opStack)
+          key = processBranchNode(key, branchNodeKey, foundNode, parentNode as TrieNode, stack)
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: The `processBranchNode` takes only 5 arguments. `opStack` was extra and was serving no purpose. Not sure why it was passed in...

File: src/baseTrie.ts [hunk 27]
```diff
@@ -669,37 +733,30 @@ module.exports = class Trie {
...
-          type: 'del',
-          key: hashRoot
-        })
-      } else {
-        opStack.push({
-          type: 'put',
-          key: hashRoot,
-          value: rlpNode
-        })
-      }
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: The reason I removed this was because I thought `BaseTrie` should not have `CheckpointTrie` specific logic. `this.isCheckpoint` is only defined in the `CheckpointTrie`. So I duplicated the `_formatNode` function in `CheckpointTrie` and moved this logic there.

File: src/baseTrie.ts [hunk 28]
```diff
@@ -740,28 +797,29 @@ module.exports = class Trie {
...
-   * @param {Function} cb
    */
-  checkRoot (root, cb) {
+  checkRoot(root: Buffer, cb: Function) {
     root = ethUtil.toBuffer(root)
-    this._lookupNode(root, (e, value) => {
+    this._lookupNode(root, (e: Error, value: TrieNode) => {
       cb(null, !!value)
     })
   }
```
[the remark this answers is not in the document]
Comment (s1na) [reply src/baseTrie.ts]: I tried to add return types for sync functions that returned something (I might have missed some though). The majority of the methods however were async and returned their values by invoking the callback.

File: src/trieNode.ts (hunk 15 above)
Comment (s1na) [reply src/trieNode.ts]: Yeah removed it.

File: src/trieNode.ts (hunk 16 above)
Comment (s1na) [reply src/trieNode.ts]: Yeah, it reads like a puzzle :) No, that's just for a typescript error, it's not in reality modifying anything.

**Comment by s1na** [comment]
coveralls error: `Bad response: 503 <h2>This website is under heavy load (queue full)</h2><p>We're sorry, too many people are accessing this website at the same time. We're working on this problem. Please try again later.</p>` :man_facepalming:

**Comment by s1na** [comment]
@holgerd77 Thanks a lot for doing this extensive review. For several comments I had to double-check to see if I had made a mistake. Glad to know if I sometime do make a mistake, you'll likely catch it!

A note on the new errors and asserts: In those places errors were possible, it's just that before the code would have failed a bit further with an unrelated error such as `undefined has no attribute X`... The typescript compiler pushed me to make the errors more explicit.

**Comment by holgerd77** [comment]
Ah, CI is not working, re-trigger didn't work.

**Comment by holgerd77** [comment]
Hmm, coverage looks good on Travis. This seems to be some (hopefully) temporary problem with Coveralls. Have temporarily disabled the check as being required and will merge now.

**Comment by s1na** [comment]
Yeah coveralls is down since yesterday (even their [homepage](https://coveralls.io/)). Hopefully it'll be fixed soon. Thanks for merging.

**Comment by holgerd77** [comment]
@s1na Just read your comment about the TS release on the other issue thread. Would it be worth to do an in-between major release here? This is hanging around for quite some time and then we would have a new work basis, eventually this would also ease the transition for people if they can do this in-between hop. Or are you planning to integrate the refactoring changes from https://github.com/ethereumjs/merkle-patricia-tree/pull/100 and subsequent PRs on a very short-term basis?

**Comment by s1na** [comment]
> Or are you planning to integrate the refactoring changes from #100 and subsequent PRs on a very short-term basis?

If we decide it's better to release all changes together I'd be okay with making the changes in a relatively short time frame.

> Would it be worth to do an in-between major release here? 

Now whether it makes sense to do a major release before making additional changes:

there have been a lot of changes that have not yet been released. Considering that doing a release is sensible. The good thing about doing a release now is that as you said the breaking changes are not that much and it'd be easier to migrate.

On the other hand, in it's current form the codebase is not very polished. I had to have a look at the recent PRs and here I wrote `I don't consider this migration complete.`. So there's definitely a lot possible changes we could do to improve things. But these improvements will likely include a few breaking changes (one major being Promises).

So in short, the current state is not so bad not to be releasable, but there is a lot of room for improvement. What do you think @holgerd77?

**Comment by holgerd77** [comment]
@s1na There not too much time pressure regarding a major release, apart from the fact that it would be nice that others can benefit of the improvements made. If you can do the simplifications/refactorings within the next 2-3 weeks I would be fine with waiting for a larger major release. We just shouldn't draw this release beyond the Christmas holidays.

**Comment by s1na** [comment]
Okay, let's target end of November as a target for release.
